pkg/solana/logpoller/loader_test.go

```diff
@@ -0,0 +1,289 @@
+	"github.com/smartcontractkit/chainlink-common/pkg/utils/tests"
+
+	"github.com/smartcontractkit/chainlink-solana/pkg/solana/logpoller"
+	mocks "github.com/smartcontractkit/chainlink-solana/pkg/solana/logpoller/mocks"
+)
+
+var (
+	messages = []string{
+		"Program J1zQwrBNBngz26jRPNWsUSZMHJwBwpkoDitXRV95LdK4 invoke [1]",
+		"Program log: Instruction: CreateLog",
+		"Program data: HDQnaQjSWwkNAAAASGVsbG8sIFdvcmxkISoAAAAAAAAA",
+		"Program J1zQwrBNBngz26jRPNWsUSZMHJwBwpkoDitXRV95LdK4 consumed 1477 of 200000 compute units",
+		"Program J1zQwrBNBngz26jRPNWsUSZMHJwBwpkoDitXRV95LdK4 success",
+	}
+)
+
+func TestEncodedLogCollector_StartClose(t *testing.T) {
+	client := new(mocks.RPCClient)
+	parser := new(mocks.ProgramEventProcessor)
+	ctx := tests.Context(t)
+
+	collector := logpoller.NewEncodedLogCollector(client, parser, logger.Nop())
+
+	assert.NoError(t, collector.Start(ctx))
+	assert.NoError(t, collector.Close())
+}
+
+func TestEncodedLogCollector_ParseSingleEvent(t *testing.T) {
+	client := new(mocks.RPCClient)
+	parser := new(testParser)
+	ctx := tests.Context(t)
+
+	collector := logpoller.NewEncodedLogCollector(client, parser, logger.Nop())
+
+	require.NoError(t, collector.Start(ctx))
+	t.Cleanup(func() {
+		require.NoError(t, collector.Close())
+	})
+
+	slot := uint64(42)
+	sig := solana.Signature{2, 1, 4, 2}
+	blockHeight := uint64(21)
+
+	client.EXPECT().GetLatestBlockhash(mock.Anything, rpc.CommitmentFinalized).Return(&rpc.GetLatestBlockhashResult{
+		RPCContext: rpc.RPCContext{
+			Context: rpc.Context{
+				Slot: slot,
+			},
+		},
+	}, nil)
+
+	client.EXPECT().GetBlocks(mock.Anything, uint64(1), mock.MatchedBy(func(val *uint64) bool {
+		return val != nil && *val == slot
+	}), mock.Anything).Return(rpc.BlocksResult{slot}, nil)
+
+	client.EXPECT().GetBlockWithOpts(mock.Anything, slot, mock.Anything).Return(&rpc.GetBlockResult{
+		Transactions: []rpc.TransactionWithMeta{
+			{
+				Meta: &rpc.TransactionMeta{
+					LogMessages: messages,
+				},
+			},
+		},
+		Signatures:  []solana.Signature{sig},
+		BlockHeight: &blockHeight,
+	}, nil).Twice()
+
+	tests.AssertEventually(t, func() bool {
+		return parser.Called()
+	})
+
+	client.AssertExpectations(t)
+}
+
+func BenchmarkEncodedLogCollector(b *testing.B) {
+	ctx := tests.Context(b)
+
+	ticker := time.NewTimer(500 * time.Millisecond)
+	defer ticker.Stop()
+
+	parser := new(testParser)
+	blockProducer := &testBlockProducer{
+		b:         b,
+		nextSlot:  10,
+		blockSigs: make(map[uint64][]solana.Signature),
+		sigs:      make(map[string]bool),
+	}
+
+	collector := logpoller.NewEncodedLogCollector(blockProducer, parser, logger.Nop())
+
+	require.NoError(b, collector.Start(ctx))
+	b.Cleanup(func() {
+		require.NoError(b, collector.Close())
+	})
+
+	b.ReportAllocs()
+	b.ResetTimer()
+
+BenchLoop:
+	for i := 0; i < b.N; i++ {
+		select {
+		case <-ticker.C:
+			blockProducer.incrementSlot()
+		case <-ctx.Done():
+			continue BenchLoop
+		default:
+			blockProducer.makeEvent()
```

Comment:
This will just emit events rapid-fire one after another with no delay in between. Is that what was intended? Would be better I think to have more control over how many events are emitted in each slot instead of leaving it up to random chance.

Reply:
The benchmark test isn't benchmarking end to end. It is only benchmarking what this component is actually doing, which is calling an RPC and parsing program logs as fast as possible with as little load on the RPC as possible. Maybe another good measurement would be the average delay between slot discovery and event 'production'. Since we are currently using polling to watch for slots, there will naturally be a delay between when a new slot is found in polling and when the log event is passed to the `ProgramEventProcessor`.

This benchmark test also does not take into account `ProgramEventProcessor` complexity/delay so some wide variance is expected, which is why it is passed around as an interface.